uc-crux-llvm/src/UCCrux/LLVM/Constraints.hs

```diff
@@ -0,0 +1,386 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators #-}
+
+module UCCrux.LLVM.Constraints
+  ( -- * Types of constraints
+    ShapeConstraint (..),
+    Constraint (..),
+    ppConstraint,
+    RelationalConstraint (..),
+
+    -- * Collections of constraints
+    Constraints (..),
+    argConstraints,
+    globalConstraints,
+    relationalConstraints,
+    emptyConstraints,
+    ppConstraints,
+    isEmpty,
+
+    -- * 'ConstrainedShape'
+    ConstrainedShape (..),
+    expand,
+    ExpansionError (..),
+    ppExpansionError,
+
+    -- * 'NewConstraint'
+    NewConstraint (..),
+    addConstraint,
+  )
+where
+
+{- ORMOLU_DISABLE -}
+import           Control.Lens (Simple, Lens, lens, (%~), (%%~), (.~), (^.))
+import           Data.Bifunctor (first)
+import           Data.BitVector.Sized (BV)
+import qualified Data.BitVector.Sized as BV
+import           Data.Coerce (coerce)
+import           Data.Function ((&))
+import           Data.Functor.Compose (Compose(..))
+import           Data.Kind (Type)
+import           Data.Map (Map)
+import qualified Data.Map as Map
+import           Data.Maybe (catMaybes)
+import           Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Vector as Vec
+import           Data.Void as Void
+import           Prettyprinter (Doc)
+import qualified Prettyprinter as PP
+
+import qualified Text.LLVM.AST as L
+
+import qualified Data.Parameterized.Context as Ctx
+import           Data.Parameterized.Classes (IxedF'(ixF'))
+import           Data.Parameterized.NatRepr (NatRepr, type (<=))
+import           Data.Parameterized.Some (Some(..))
+import           Data.Parameterized.TraversableFC (allFC, fmapFC, toListFC)
+
+import           Lang.Crucible.LLVM.DataLayout (Alignment, fromAlignment)
+
+import           UCCrux.LLVM.Cursor (Selector(..), SomeInSelector(SomeInSelector), seekType)
+import           UCCrux.LLVM.Errors.Unimplemented (unimplemented)
+import qualified UCCrux.LLVM.Errors.Unimplemented as Unimplemented
+import           UCCrux.LLVM.Shape (Shape, ShapeSeekError)
+import qualified UCCrux.LLVM.Shape as Shape
+import           UCCrux.LLVM.FullType.ModuleTypes (ModuleTypes)
+import           UCCrux.LLVM.FullType.MemType (asFullType)
+import           UCCrux.LLVM.FullType.Type (FullType(..), FullTypeRepr(FTPtrRepr))
+{- ORMOLU_ENABLE -}
+
+--------------------------------------------------------------------------------
+
+-- * Types of constraints
+
+data ShapeConstraint (m :: Type) (atTy :: FullType m) where
+  -- | This pointer is not null, and has space for at least this many elements
+  --
+  -- Invariant: argument should be positive.
+  Allocated :: !Int -> ShapeConstraint m ('FTPtr ft)
+  -- | This pointer points to initialized memory with at least this many elements
+  --
+  -- Invariant: argument should be positive.
+  Initialized :: !Int -> ShapeConstraint m ('FTPtr ft)
```

Comment:
Unless there's a very good reason, I've started trying hard to avoid the limited precision types.  We've had to diagnose too many bugs that were eventually traced to overflow problems to make me comfortable with them, so I'd vote for `Integer` or `Natural` here.

Reply:
> Unless there's a very good reason, I've started trying hard to avoid the limited precision types. We've had to diagnose too many bugs that were eventually traced to overflow problems to make me comfortable with them, so I'd vote for Integer or Natural here.

Huh, that's interesting! Good to know, I'll note in the follow-up to prefer `Natural`.

Reply:
https://github.com/GaloisInc/crucible/issues/692